### PR TITLE
[DNM] use bindgen to auto generate ffi rs code

### PR DESCRIPTION
**Pull request by fredchenbj**
just use old c.h and bindgen to gen ffi rs code. and fix to pass `cargo build.`

this is step 1, later replace with rocksdb's c.h to gen code.

Signed-off-by: fredchenbj <cfworking@163.com>

### CODE REVIEW

**Comment by Connor1996** [comment]
Please fix the conflicts

**Comment by fredchenbj** [comment]
> Please fix the conflicts

OK, I will fix ASAP.

File: librocksdb_sys/crocksdb/crocksdb/c.h [hunk 1]
```diff
@@ -940,10 +943,10 @@ extern C_ROCKSDB_LIBRARY_API void crocksdb_options_enable_statistics(
     crocksdb_options_t*, unsigned char);
 extern C_ROCKSDB_LIBRARY_API void crocksdb_options_reset_statistics(
     crocksdb_options_t*);
-extern C_ROCKSDB_LIBRARY_API bool crocksdb_load_latest_options(
+extern C_ROCKSDB_LIBRARY_API unsigned char crocksdb_load_latest_options(
```
Comment (Connor1996) [inline librocksdb_sys/crocksdb/crocksdb/c.h]: `bindgen` generates FFI based on `c.h`, but why we modify `c.h`?

Comment (fredchenbj) [reply librocksdb_sys/crocksdb/crocksdb/c.h]: @Connor1996 this function was added by us, so not conform to c's style.

Comment (fredchenbj) [reply librocksdb_sys/crocksdb/crocksdb/c.h]: rocksdb's c.h has convention: 
`Bools have the type unsigned char (0 == false; rest == true)`

Comment (Connor1996) [reply librocksdb_sys/crocksdb/crocksdb/c.h]: got it

**Comment by yiwu-arbug** [comment]
@fredchenbj awesome changes! Some general questions:

1. Can you summarize what's changed in this PR? Seems mostly typing changes. Why are they needed?
2. The PR seems not including the generated code. Should we checkin the generated code or not?
3. What's the overall plan for migration?

Thanks.

**Comment by BusyJay** [comment]
It introduces clang dependency, which can be a pain to fix. You can learn how to avoid it from tikv/grpc-rs.

**Comment by Connor1996** [comment]
@BusyJay You mean pre-generate binding? Yeah, that's what we plan to do. For easy reviewing, we plan to do it step by step.

File: src/event_listener.rs [hunk 2]
```diff
@@ -155,17 +153,16 @@ impl IngestionInfo {
...
-#[repr(transparent)]
-pub struct WriteStallInfo(DBWriteStallInfo);
+pub struct WriteStallInfo(crocksdb_writestallinfo_t);
 
 impl WriteStallInfo {
     pub fn cf_name(&self) -> &str {
         unsafe { fetch_str!(crocksdb_writestallinfo_cf_name(&self.0)) }
     }
-    pub fn cur(&self) -> WriteStallCondition {
+    pub fn cur(&self) -> crocksdb_writestallcondition_t {
```
Comment (Connor1996) [inline src/event_listener.rs]: All public functions/traits of `src/xxx.rs` should be unchanged, otherwise `TiKV` or other users may change code dramatically. Consider wrapping `crocksdb_writestallcondition_t` with `WriteStallCondition`, just like how `WriteStallInfo` does. Same for other types.

File: librocksdb_sys/crocksdb/crocksdb/c.h (hunk 1 above)
Comment (Connor1996) [reply librocksdb_sys/crocksdb/crocksdb/c.h]: can you extract them to a separate PR?

File: src/table_properties_collector.rs [hunk 3]
```diff
@@ -26,7 +28,7 @@ use std::slice;
 /// TablePropertiesCollector object per table and then call it sequentially
 pub trait TablePropertiesCollector {
     /// Will be called when a new key/value pair is inserted into the table.
-    fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, seq: u64, file_size: u64);
+    fn add(&mut self, key: &[i8], value: &[i8], entry_type: DBEntryType, seq: u64, file_size: u64);
```
[the remark this answers is not in the document]
Comment (fredchenbj) [reply src/table_properties_collector.rs]: for `const char*` in c.h is transferred to `*const libc::c_char` by bindgen, which required `i8`.  All transfer like this is for this. @Connor1996

File: src/event_listener.rs (hunk 2 above)
Comment (fredchenbj) [reply src/event_listener.rs]: OK

**Comment by fredchenbj** [comment]
> @fredchenbj awesome changes! Some general questions:
> 
> 1. Can you summarize what's changed in this PR? Seems mostly typing changes. Why are they needed?
> 2. The PR seems not including the generated code. Should we checkin the generated code or not?
> 3. What's the overall plan for migration?
> 
> Thanks.

@yiwu-arbug Thanks for your questions. 

for 3, the migration plan has three steps: 1) use the old crocksdb's c.h and bindgen to generate ffi rust code, and make it work for all dependency code; 2) replace with rocksdb's c.h, and make all code in c.h conform to rocksdb's style, here we would only include rocksdb's c.h and not change it at all; 3) try to use rocksdb cpp api to generate ffi rust code.

for 1, this pr has mainly three changes: 1) change crocksdb's c.h and c.cc to conform to bindgen's requirement; 2) use bindgen to generate ffi rust code; 3) change code in rust-rocksdb, including many type changes, to make those code could call the code geneated by bindgen.

for 2, the generated code is in `target/debug/build/librocksdb_sys-f21a7cfdab79a481/out/bindings.rs`, which is auto generated by `cargo build`. IMHO, we don't need checkin it.

File: librocksdb_sys/crocksdb/crocksdb/c.h (hunk 1 above)
Comment (fredchenbj) [reply librocksdb_sys/crocksdb/crocksdb/c.h]: OK

File: src/table_properties_collector.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/table_properties_collector.rs]: see reference
- rust std implementation https://doc.rust-lang.org/src/std/ffi/c_str.rs.html#975,
- https://doc.rust-lang.org/std/os/raw/type.c_char.html, 
- https://stackoverflow.com/questions/47684111/make-string-type-compatible-with-arm/47684200. 
`i8` or `u8` is only due to how C `char` represents in different platforms. But the data is always one-byte long. As `char` in Rust is represented as `u8`, so we must convert it to `u8`.

[the remark this answers is not in the document]
Comment (fredchenbj) [reply src/table_properties_collector.rs]: got it. I would fix it.

**Comment by yiwu-arbug** [comment]
@fredchenbj for 3, can you give more details for step 1)? Does the generated code provide difference interface than the existing rust API? If so, how much work is it to update tikv? If that work is huge, is there way to split the work into small bits?

Also for 3, if we will trying generating code from c++ directly, how useful step 2) would be?

Currently the wrapper doesn't handle `Status` return from rocksdb well. It convert the struct to string and miss the typing. I hope with step 3) we can change it to convert to rust `Result`.
